compiler/rustc_symbol_mangling/src/v0.rs

```diff
@@ -485,27 +485,36 @@ impl Printer<'tcx> for SymbolMangler<'tcx> {
         mut self,
         predicates: &'tcx ty::List<ty::Binder<'tcx, ty::ExistentialPredicate<'tcx>>>,
     ) -> Result<Self::DynExistential, Self::Error> {
+        let mut last_trait_ref = None;
+
         for predicate in predicates {
-            self = self.in_binder(&predicate, |mut cx, predicate| {
-                match predicate {
-                    ty::ExistentialPredicate::Trait(trait_ref) => {
+            match predicate.as_ref().skip_binder() {
+                ty::ExistentialPredicate::Trait(trait_ref) => {
+                    last_trait_ref = Some(predicate.rebind(*trait_ref));
+                    // We only use `in_binder` on trait predicates due to issue #83611.
+                    self = self.in_binder(&predicate, |mut cx, _predicate| {
                         // Use a type that can't appear in defaults of type parameters.
                         let dummy_self = cx.tcx.mk_ty_infer(ty::FreshTy(0));
                         let trait_ref = trait_ref.with_self_ty(cx.tcx, dummy_self);
                         cx = cx.print_def_path(trait_ref.def_id, trait_ref.substs)?;
-                    }
-                    ty::ExistentialPredicate::Projection(projection) => {
-                        let name = cx.tcx.associated_item(projection.item_def_id).ident;
-                        cx.push("p");
-                        cx.push_ident(&name.as_str());
-                        cx = projection.ty.print(cx)?;
-                    }
-                    ty::ExistentialPredicate::AutoTrait(def_id) => {
-                        cx = cx.print_def_path(*def_id, &[])?;
-                    }
+                        Ok(cx)
+                    })?;
                 }
-                Ok(cx)
-            })?;
+                ty::ExistentialPredicate::Projection(projection) => {
+                    let last_trait_ref = last_trait_ref
+                        .expect("trait predicate must come before projection predicate");
+                    assert_eq!(last_trait_ref.bound_vars(), predicate.bound_vars());
```

Comment:
This assertion probably won't actually hold when there are supertrait HRTBs.
Ex.
```
trait Foo<'c> { type Output; }
trait Bar<'a, 'b>: for<'c> Foo<'c> { }

dyn for<'a, 'b> Bar<Output = u32>
````